### ISSUE TITLE
No render for windows 3dviewer's built-in Bee.glb

### ISSUE THREAD

**Issue by HO-COOH** [f3d-app/f3d#943]
**Describe the bug**
It seems it does not able to render that file, no matter how I zoom in or zoom out.

**To Reproduce**
1. Export that file
![image](https://github.com/f3d-app/f3d/assets/42881734/635b1f6e-4ca6-4216-b735-ca166319557b)
2. Drag it into the latest release of f3d.
![image](https://github.com/f3d-app/f3d/assets/42881734/addb970c-c144-4c63-b73f-23e36c6b093e)

**Expected behavior**
Renders like 3dviewer.

**System Information:**
 - OS: Windows 11
 - GPU and GPU driver: RTX3060

**F3D Information**
Paste the content of `f3d --version`:
![image](https://github.com/f3d-app/f3d/assets/42881734/e22f778f-92d0-474a-916e-a4be4bf3850f)

**Additional context**

**Comment by mwestphal** [comment]
Can you share bee.glb ?

**Comment by HO-COOH** [comment]
Sure. [Bee.zip](https://github.com/f3d-app/f3d/files/12305540/Bee.zip)

**Comment by HO-COOH** [comment]
[Windmill.zip](https://github.com/f3d-app/f3d/files/12311347/Windmill.zip)

Seems like yes. Just tried several others, all failed to load. Is there any kind of error reported in f3d?

**Comment by mwestphal** [comment]
It doesn't, it fails inside the document loader. No idea why for now.

**Comment by mwestphal** [comment]
Sorry, it took me a while to check this one. Your file and others from windows 3D viewer are using a glTF extension that is not supported by VTK:

```
ERROR: In vtkGLTFDocumentLoaderInternals.cxx, line 1305
vtkGLTFDocumentLoader (0x55e7bbe62820): glTF extension KHR_materials_pbrSpecularGlossiness is required in this model, but not supported by this loader. Aborting
```

I see two issues here:

1. VTK should support this extension as it seems an important one
2. It would be better if F3D gave a better feedback in that case except for an empty window

**Comment by HO-COOH** [comment]
> Sorry, it took me a while to check this one. Your file and others from windows 3D viewer are using a glTF extension that is not supported by VTK:
> 
> ```
> ERROR: In vtkGLTFDocumentLoaderInternals.cxx, line 1305
> vtkGLTFDocumentLoader (0x55e7bbe62820): glTF extension KHR_materials_pbrSpecularGlossiness is required in this model, but not supported by this loader. Aborting
> ```
> 
> I see two issues here:
> 
> 1. VTK should support this extension as it seems an important one
> 2. It would be better if F3D gave a better feedback in that case except for an empty window

Thanks for finding it out. And yes, please add this log to the message window otherwise users would think it loads successfully (because the file name is already shown in the rendering window) and find it confusing.

**Comment by mwestphal** [comment]
> Thanks for finding it out. And yes, please add this log to the message window otherwise users would think it loads 
> successfully (because the file name is already shown in the rendering window) and find it confusing.

I agree, currently you would need a debug version to get this output, which is not ideal.

**Comment by mwestphal** [comment]
related to https://github.com/f3d-app/f3d/issues/653

**Comment by mwestphal** [comment]
renamed this one and opened another one for the debug output: https://github.com/f3d-app/f3d/issues/966

**Comment by mwestphal** [comment]
> Thanks for finding it out. And yes, please add this log to the message window otherwise users would think it loads successfully (because the file name is already shown in the rendering window) and find it confusing.

The log issue was fixed

**Comment by cryptosebek** [comment]
I utilize f3d to create miniature model thumbnails (128x128 pixels). It's essential for my purpose to render the model with a degree of approximation whenever feasible, rather than experiencing a complete failure due to an error.

For anybody struggling with this issue, here's a snippet I wrote for Node.JS, which essentially just replacing "KHR_materials_pbrSpecularGlossiness" with "pbrMetallicRoughness" as a fallback:
```
const fs = require("fs").promises;

async function readGLB(filePath) {
  const data = await fs.readFile(filePath);
  const header = {
    magic: data.readUInt32LE(0),
    version: data.readUInt32LE(4),
    length: data.readUInt32LE(8),
  };

  if (header.magic !== 0x46546c67) {
    throw new Error("File is not a valid GLB format");
  }
  let offset = 12;
  let jsonChunkLength = data.readUInt32LE(offset);
  let jsonChunkType = data.readUInt32LE(offset + 4);
  if (jsonChunkType !== 0x4e4f534a) {
    throw new Error("First chunk is not JSON");
  }
  const jsonChunk = data.slice(offset + 8, offset + 8 + jsonChunkLength);
  const json = JSON.parse(jsonChunk.toString());
  return { json, binaryData: data, jsonChunkLength, offset };
}

async function removeMaterials(glbData) {
  const { json } = glbData;
  delete json.extensionsUsed;
  delete json.extensionsRequired;
  json.materials.forEach((material) => {
    if (
      material.extensions &&
      material.extensions.KHR_materials_pbrSpecularGlossiness
    ) {
      material.pbrMetallicRoughness = {
        baseColorFactor:
          material.extensions.KHR_materials_pbrSpecularGlossiness.diffuseFactor,
        metallicFactor:
          material.extensions.KHR_materials_pbrSpecularGlossiness
            .glossinessFactor,
        roughnessFactor:
          material.extensions.KHR_materials_pbrSpecularGlossiness
            .glossinessFactor,
        baseColorTexture:
          material.extensions.KHR_materials_pbrSpecularGlossiness
            .diffuseTexture,
      };
      delete material.extensions;
    }
  });
  return { ...glbData, json };
}

async function writeGLB(modifiedGLB, outputFilePath) {
  const { json, binaryData, jsonChunkLength, offset } = modifiedGLB;
  let newJsonChunk = Buffer.from(JSON.stringify(json));
  newJsonChunk = Buffer.concat([
    newJsonChunk,
    Buffer.alloc(jsonChunkLength - newJsonChunk.length, 0x20),
  ]);
  binaryData.fill(newJsonChunk, offset + 8, offset + 8 + jsonChunkLength);
  await fs.writeFile(outputFilePath, binaryData);
}

async function patchGlb(path) {
  try {
    const glbData = await readGLB(path);
    if (
      !glbData.json ||
      !glbData.json.extensionsUsed ||
      !glbData.json.extensionsUsed.includes(
        "KHR_materials_pbrSpecularGlossiness"
      )
    ) {
      return;
    }
    const modifiedGLB = await removeMaterials(glbData);
    await writeGLB(modifiedGLB, path);
  } catch (error) {
    console.error("Error:", error);
  }
}

module.exports = { patchGlb };
```

![image](https://github.com/f3d-app/f3d/assets/93789460/0dfe9609-de17-401b-96a6-cd991f8a71c4)
